File: extension/src/types/analysis.ts
export type RiskLabel = "safe" | "suspicious" | "dangerous";
export type AnalysisMode = "checking" | "local-only" | "backend-enriched" | "backend-unavailable" | "cached";

export interface DOMFeatures {
  has_password_field: boolean;
  num_forms: number;
  external_form_action: boolean;
  num_iframes: number;
  external_links_ratio: number;
  has_hidden_inputs: boolean;
}

export interface URLFeatures {
  url_length: number;
  num_dots: number;
  num_hyphens: number;
  uses_ip_domain: boolean;
  has_at_symbol: boolean;
  uses_https: boolean;
  num_subdomains: number;
  suspicious_keywords: string[];
  uses_punycode: boolean;
  domain_entropy: number;
  domain: string;
  typosquat_target: string | null;
  typosquat_distance: number | null;
  typosquat_is_homograph: boolean;
  mixed_script_label: boolean;
}

export interface AnalysisSources {
  heuristics: boolean;
  ml: boolean;
  phishtank: boolean;
  tls: boolean;
  demo?: boolean;
}

export type RiskBreakdownCategory = "url" | "dom" | "threat_intel" | "tls" | "ml";

export interface RiskBreakdownItem {
  category: RiskBreakdownCategory;
  score: number;
  min_score: number;
  max_score: number;
  reasons: string[];
  source: string;
}

export interface AnalysisResponse {
  risk_score: number;
  label: RiskLabel;
  confidence: number;
  reasons: string[];
  sources: AnalysisSources;
  risk_breakdown?: RiskBreakdownItem[];
}

export interface PopupAnalysis extends AnalysisResponse {
  url: string;
  backendAvailable: boolean;
  mode: AnalysisMode;
  analyzedAt: string;
}

export interface ExtensionSettings {
  backendBaseUrl: string;
  requestTimeoutMs: number;
  dangerOverlayEnabled: boolean;
}

export interface FeedbackReport {
  url: string;
  observed_label: RiskLabel;
  expected_label: RiskLabel;
  notes?: string;
}

export interface BackendHealthResponse {
  status: "ok" | string;
  service: string;
}

export interface DiagnosticsCapabilities {
  diagnostics_enabled: boolean;
  rate_limiting_enabled: boolean;
  threat_intel_enabled: boolean;
  tls_analysis_enabled: boolean;
  ml_model_available: boolean;
  demo_threat_source_enabled: boolean;
}

export interface DiagnosticsResponse {
  status: "ok" | string;
  service: string;
  privacy: string;
  capabilities: DiagnosticsCapabilities;
  counters: Record<string, number>;
  labels: Record<string, number>;
  sources: Record<string, number>;
}

export type BackendStatusState = "online" | "diagnostics-disabled" | "offline";

export interface BackendStatus {
  state: BackendStatusState;
  service: string | null;
  diagnosticsAvailable: boolean;
  diagnostics: DiagnosticsResponse | null;
  message: string;
  checkedAt: string;
}

File: backend/app/schemas/analysis.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.services.url_normalizer import URLNormalizationError, normalize_url


RiskLabel = Literal["safe", "suspicious", "dangerous"]
RiskCategory = Literal["url", "dom", "threat_intel", "tls", "domain_age", "ml"]


class DOMFeatures(BaseModel):
    has_password_field: bool = False
    num_forms: int = Field(default=0, ge=0, le=100)
    external_form_action: bool = False
    num_iframes: int = Field(default=0, ge=0, le=200)
    external_links_ratio: float = Field(default=0.0, ge=0.0, le=1.0)
    has_hidden_inputs: bool = False


class AnalysisRequest(BaseModel):
    url: str = Field(min_length=8, max_length=2048)
    dom_features: DOMFeatures = Field(default_factory=DOMFeatures)

    @field_validator("url")
    @classmethod
    def validate_http_url(cls, value: str) -> str:
        try:
            return normalize_url(value)
        except URLNormalizationError as exc:
            raise ValueError("url must be an absolute http or https URL") from exc


class AnalysisSources(BaseModel):
    heuristics: bool
    ml: bool
    phishtank: bool
    tls: bool
    domain_age: bool = False
    demo: bool = False


class RiskBreakdownItem(BaseModel):
    category: RiskCategory
    score: int = Field(ge=-10, le=100)
    min_score: int = Field(default=0, ge=-10, le=100)
    max_score: int = Field(ge=0, le=100)
    reasons: list[str]
    source: str


class AnalysisResponse(BaseModel):
    risk_score: int = Field(ge=0, le=100)
    label: RiskLabel
    confidence: float = Field(ge=0.0, le=1.0)
    reasons: list[str]
    sources: AnalysisSources
    risk_breakdown: list[RiskBreakdownItem] = Field(default_factory=list)


class ReportRequest(BaseModel):
    url: str = Field(min_length=8, max_length=2048)
    observed_label: RiskLabel
    expected_label: RiskLabel
    notes: str | None = Field(default=None, max_length=500)

    @field_validator("url")
    @classmethod
    def validate_http_url(cls, value: str) -> str:
        try:
            return normalize_url(value)
        except URLNormalizationError as exc:
            raise ValueError("url must be an absolute http or https URL") from exc


class ReportResponse(BaseModel):
    status: Literal["received"]
    message: str
